File: Longest Valid Parentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        Stack<Integer> left = new Stack<Integer>();
        int right = -1;
        int maxLen = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                left.push(i);
            }
            else {
                if (left.empty()) {
                    right = i;
                }
                else {
                    left.pop();
                    int start = left.empty()?right:left.peek();
                    int len = i - start;
                    if (len > maxLen)
                        maxLen = len;
                }
            }
        }
        return maxLen;
    }
}

File: Longest Valid Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> left;
        int right = -1;
        int max_len = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                // left
                left.push(i);
            }
            else {
                // right
                if (left.empty()) {
                    right = i;
                }
                else {
                    left.pop();
                    int start = left.empty()?right:left.top();
                    int len = i - start;
                    if (len > max_len)
                        max_len = len;
                }
            }
        }
        return max_len;
    }
};